connectors/amazons3/connector/src/main/java/org/apache/manifoldcf/crawler/connectors/amazons3/GenericDocumentProcess.java

```diff
@@ -0,0 +1,276 @@
+ * the License.
+ */
+package org.apache.manifoldcf.crawler.connectors.amazons3;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InterruptedIOException;
+import java.util.Date;
+import java.util.Set;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
+import org.apache.manifoldcf.agents.interfaces.RepositoryDocument;
+import org.apache.manifoldcf.agents.interfaces.ServiceInterruption;
+import org.apache.manifoldcf.amazons3.S3Artifact;
+import org.apache.manifoldcf.core.interfaces.ManifoldCFException;
+import org.apache.manifoldcf.core.interfaces.Specification;
+import org.apache.manifoldcf.crawler.interfaces.IExistingVersions;
+import org.apache.manifoldcf.crawler.interfaces.IProcessActivity;
+import org.apache.manifoldcf.crawler.system.Logging;
+
+import com.amazonaws.AmazonClientException;
+import com.amazonaws.AmazonServiceException;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.AccessControlList;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.Grant;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.S3Object;
+
+/**
+ * Generic amazons3 extractor
+ * @author Kuhajeyan
+ *
+ */
+public class GenericDocumentProcess extends AmazonS3DocumentProcessUtility
+    implements DocumentProcess {
+
+  private static final String TEXT_PLAIN = "text/plain";
+
+  /**
+   * Process documents with out any tika extractor
+   * @param documentIdentifiers
+   * @param statuses
+   * @param spec
+   * @param activities
+   * @param jobMode
+   * @param usesDefaultAuthority
+   * @param amazons3Client
+   * @throws ManifoldCFException
+   */
+  @Override
+  public void doProcessDocument(String[] documentIdentifiers,
+      IExistingVersions statuses, Specification spec,
+      IProcessActivity activities, int jobMode,
+      boolean usesDefaultAuthority, AmazonS3 amazons3Client)
+      throws ManifoldCFException, ServiceInterruption {
+    if (amazons3Client == null)
+      throw new ManifoldCFException(
+          "Amazon client can not connect at the moment");
+
+    for (String documentIdentifier : documentIdentifiers) {
+      try {
+        if (documentIdentifier == null
+            || StringUtils.isEmpty(documentIdentifier)) {
+          Logging.connectors
+              .warn("Document identifier is empty, document will not be processed");
+          continue;
+        }
+
+        String versionString;
+        String[] aclsToUse;
+
+        if (documentIdentifier
+            .split(AmazonS3Config.STD_SEPARATOR_BUCKET_AND_KEY) == null
+            && documentIdentifier.length() < 1) {
+          continue;
+        }
+
+        S3Artifact s3Artifact = getS3Artifact(documentIdentifier);
+        S3Object s3Obj = amazons3Client.getObject(new GetObjectRequest(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *s3Obj*. Currently, there are execution paths that do not contain closure statements. Either a) close *s3Obj* in a try-finally block or b) close the resource by declaring *s3Obj* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).